### PR TITLE
refactor(rest): use dynamic value provider for actions

### PR DESCRIPTION
**Pull request by raymondfeng**
Optimize actions using static providers.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by bajtos** [comment]
I don't have strong opinions on the actual change from instance providers to static providers, I'll leave that up to code owners to decide (@hacksparrow @emonddr @deepakrkris).

**Comment by raymondfeng** [comment]
@emonddr See https://github.com/strongloop/loopback-next/pull/5370

**Comment by raymondfeng** [comment]
> Have you considered preserving current instance-based providers as thin wrappers around the new static providers for backwards compatibility? Maybe it's not worth the effort, let's discuss.

That's probably not worth the effort. Too many choices are bad :-).

**Comment by bajtos** [comment]
> > Have you considered preserving current instance-based providers as thin wrappers around the new static providers for backwards compatibility? Maybe it's not worth the effort, let's discuss.
> 
> That's probably not worth the effort. Too many choices are bad :-).

That's true :+1:

Please update the commit message to communicate this change as backwards-incompatible.

**Comment by raymondfeng** [comment]
@jannyHou @hacksparrow Please review.

File: packages/rest/src/providers/send.provider.ts [hunk 1]
```diff
@@ -16,9 +15,10 @@ import {writeResultToResponse} from '../writer';
  * @returns The handler function that will populate the
  * response with operation results.
  */
+
 @injectable({scope: BindingScope.SINGLETON})
-export class SendProvider implements Provider<Send> {
-  value() {
+export class SendProvider {
+  static value() {
```
Comment (jannyHou) [inline packages/rest/src/providers/send.provider.ts:21]: not necessarily in this PR, can we also turn `SendResponseMiddlewareProvider` into a static Provider?

Comment (raymondfeng) [reply packages/rest/src/providers/send.provider.ts:21]: It's not needed as `SendResponseMiddlewareProvider` is a singleton.

File: packages/rest/src/providers/invoke-method.provider.ts [hunk 2]
```diff
@@ -15,19 +15,17 @@ import debugFactory from 'debug';
...
-    return (route, args) => this.action(route, args);
-  }
-
-  action(route: RouteEntry, args: OperationArgs): Promise<OperationRetval> {
-    return route.invokeHandler(this.context, args);
+export class InvokeMethodProvider {
+  static value(
+    @inject(RestBindings.Http.CONTEXT) context: Context,
+  ): InvokeMethod {
+    const invokeMethod: InvokeMethod = (route, args) =>
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/rest/src/providers/invoke-method.provider.ts:26]: They are inferred from `InvokeMethod` type.

File: packages/rest/src/providers/find-route.provider.ts [hunk 3]
```diff
@@ -14,27 +14,28 @@ import {asMiddleware, Middleware} from '@loopback/express';
...
-    const found = this.handler.findRoute(request);
-    debug('Route found for %s %s', request.method, request.originalUrl, found);
-    found.updateBindings(this.context);
-    return found;
+export class FindRouteProvider {
+  static value(
+    @inject(RestBindings.Http.CONTEXT) context: Context,
+    @inject(RestBindings.HANDLER) handler: HttpHandler,
+  ): FindRoute {
+    const findRoute: FindRoute = request => {
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/rest/src/providers/find-route.provider.ts:27]: It's inferred from `FindRoute`.

**Comment by raymondfeng** [comment]
> According to the benchmark this PR would have a significant improvement for request processing

This one is minor improvement.

**Pull request merged**